### PR TITLE
utils_net: verify_ip_address_ownership() on remote host if VM is migrated

### PR DESCRIPTION
**Pull request by balamuruhans**
`verify_ip_address_ownership()` fails when ip address of VM is queried
after it is migrated to target host and cause the tests to error, handle
it using session to check on the remote host.

Signed-off-by: Balamuruhan S <bala24@linux.vnet.ibm.com>

### CODE REVIEW

File: virttest/utils_net.py [hunk 1]
```diff
@@ -1720,18 +1723,21 @@ def refresh_neigh_table(interface_name=None, neigh_address="ff02::1"):
...
-    refresh_neigh_table(interface_name, neigh_address)
+    refresh_neigh_table(interface_name, neigh_address, session=session)
+    func = process.getoutput
+    if session:
+        func = session.cmd_output
     cmd = "ip -6 neigh show nud reachable"
     if neigh_address:
         cmd += " %s" % neigh_address
-    output = decode_to_text(process.system_output(cmd))
+    output = func(cmd)
```
Comment (vivianQizhu) [inline virttest/utils_net.py]: Seems this way you loose `decode()`, and also `process.system_output` only returns `cmd_result.stdout`,  but `process.getoutput` returns `cmd_result.exit_status, cmd_result.stdout_text`, possibly L#1745 - L#1752 wont work properly.

Comment (balamuruhans) [reply virttest/utils_net.py]: @vivianQizhu I couldn't get you, please correct me if I am wrong,
`process.getoutput()` returns `cmd_result.stdout_text` and property `stdout_text` takes care of decode part as well which returns `stdout`. how does this affect lines that is mentioned ?

Comment (vivianQizhu) [reply virttest/utils_net.py]: @balamuruhans You are right. Sorry that I didn't notice the difference between `stdout_text` and `stdout`, and also missed the `[1]` in `getoutput()` return line.

**Comment by balamuruhans** [comment]
Tested it and it is working fine,

**Issue observed:**
```
2019-03-31 07:18:09,989 virt_vm          L1347 DEBUG| Attempting to log into 'vms3' via serial console (timeout 240s)
2019-03-31 07:18:10,490 remote           L0212 DEBUG| Got console prompt, send return to show login
2019-03-31 07:18:12,169 client           L1126 DEBUG| Sending command: cat /proc/uptime
2019-03-31 07:18:12,276 virsh_migrate_st L0050 INFO | uptime of migrated VM vms3: 476.29
2019-03-31 07:18:12,988 virsh_migrate_st L0057 INFO | Guest state is 'running' at destination is as expected
2019-03-31 07:18:12,988 remote           L0395 DEBUG| Attempting to log into 10.20.19.21:22 using ssh (timeout 240s)
2019-03-31 07:18:35,472 ip_sniffing      L0068 DEBUG| Dropped the address cache of HWADDR (52:54:00:9b:9c:9d)
2019-03-31 07:18:35,473 virsh_migrate_st L0237 DEBUG| Cleanup vms...

```
**After fix:**
```
2019-04-01 07:58:57,911 virt_vm          L1346 DEBUG| Attempting to log into 'vms3' via serial console (timeout 240s)
2019-04-01 07:58:58,412 remote           L0212 DEBUG| Got console prompt, send return to show login
2019-04-01 07:59:00,085 client           L1126 DEBUG| Sending command: cat /proc/uptime
2019-04-01 07:59:00,191 virsh_migrate_st L0050 INFO | uptime of migrated VM vms3: 531.8
2019-04-01 07:59:00,916 virsh_migrate_st L0057 INFO | Guest state is 'running' at destination is as expected
2019-04-01 07:59:00,917 remote           L0395 DEBUG| Attempting to log into 10.20.19.21:22 using ssh (timeout 240s)
2019-04-01 07:59:02,302 client           L1126 DEBUG| Sending command: cat /proc/net/arp
2019-04-01 07:59:02,406 libvirt          L1252 INFO | Check VM network connectivity after migrating
2019-04-01 07:59:02,406 client           L1162 DEBUG| Sending command (safe): echo %OS%
2019-04-01 07:59:02,508 client           L1126 DEBUG| Sending command: ping 192.168.122.81  -c 10
2019-04-01 07:59:02,612 client           L0852 DEBUG| PING 192.168.122.81 (192.168.122.81) 56(84) bytes of data.
2019-04-01 07:59:02,612 client           L0852 DEBUG| 64 bytes from 192.168.122.81: icmp_seq=1 ttl=64 time=0.110 ms
2019-04-01 07:59:03,652 client           L0852 DEBUG| 64 bytes from 192.168.122.81: icmp_seq=2 ttl=64 time=0.085 ms
2019-04-01 07:59:04,692 client           L0852 DEBUG| 64 bytes from 192.168.122.81: icmp_seq=3 ttl=64 time=0.070 ms
2019-04-01 07:59:05,732 client           L0852 DEBUG| 64 bytes from 192.168.122.81: icmp_seq=4 ttl=64 time=0.070 ms
2019-04-01 07:59:06,771 client           L0852 DEBUG| 64 bytes from 192.168.122.81: icmp_seq=5 ttl=64 time=0.068 ms
2019-04-01 07:59:07,811 client           L0852 DEBUG| 64 bytes from 192.168.122.81: icmp_seq=6 ttl=64 time=0.070 ms
2019-04-01 07:59:08,850 client           L0852 DEBUG| 64 bytes from 192.168.122.81: icmp_seq=7 ttl=64 time=0.066 ms
2019-04-01 07:59:09,891 client           L0852 DEBUG| 64 bytes from 192.168.122.81: icmp_seq=8 ttl=64 time=0.066 ms
2019-04-01 07:59:10,931 client           L0852 DEBUG| 64 bytes from 192.168.122.81: icmp_seq=9 ttl=64 time=0.066 ms
2019-04-01 07:59:11,971 client           L0852 DEBUG| 64 bytes from 192.168.122.81: icmp_seq=10 ttl=64 time=0.066 ms
2019-04-01 07:59:11,971 client           L0852 DEBUG| 
2019-04-01 07:59:11,972 client           L0852 DEBUG| --- 192.168.122.81 ping statistics ---
2019-04-01 07:59:11,972 client           L0852 DEBUG| 10 packets transmitted, 10 received, 0% packet loss, time 9359ms
2019-04-01 07:59:11,972 client           L0852 DEBUG| rtt min/avg/max/mdev = 0.066/0.073/0.110/0.016 ms

```

File: virttest/utils_net.py [hunk 2]
```diff
@@ -1765,11 +1771,11 @@ def get_neigh_attch_interface(neigh_address):
     return get_neighbours_info(neigh_address)[neigh_address]["attach_if"]
```
Comment (vivianQizhu) [inline virttest/utils_net.py]: Do we need support `session` here? And would you help correct the typo in doc string: `wihch`?

File: virttest/utils_net.py [hunk 3]
```diff
@@ -3087,38 +3096,43 @@ def verify_ip_address_ownership(ip, macs, timeout=20.0, devs=None):
...
             arping_cmd = "%s -C1 -c3 -w%d -I %s %s" % (arping_bin, int(timeout),
                                                        dev, ip)
         else:
             arping_cmd = "%s -f -c3 -w%d -I %s %s" % (arping_bin, int(timeout),
                                                       dev, ip)
         try:
-            o = decode_to_text(process.system_output(arping_cmd, shell=True, verbose=False))
-        except process.CmdError:
+            o = func(arping_cmd)
+        except (process.CmdError, aexpect.ShellError, aexpect.ShellTimeoutError):
```
Comment (vivianQizhu) [inline virttest/utils_net.py]: Here `aexpect.ShellError` is enough, it should cover `aexpect.ShellTimeoutError` and all other sub-type.

File: virttest/utils_net.py (hunk 2 above)
Comment (balamuruhans) [reply virttest/utils_net.py]: okay, currently it is not necessary but it is right way to add `session` support for `neigh_reachable()` and `get_neigh_attch_interface()` which depends on `get_neighbours_info()`. I will fix the typo too.

File: virttest/utils_net.py (hunk 3 above)
Comment (balamuruhans) [reply virttest/utils_net.py]: cool, I will make the suggested change.

**Comment by balamuruhans** [comment]
@vivianQizhu taken care of the suggested changes.

File: virttest/utils_net.py [hunk 4]
```diff
@@ -1747,29 +1753,29 @@ def get_neighbours_info(neigh_address="", interface_name=None):
...
 
 
-def get_neigh_attch_interface(neigh_address):
+def get_neigh_attch_interface(neigh_address, session=None):
     """
-    Get the interface wihch can reach the neigh_address
+    Get the interface which can reach the neigh_address
     """
-    return get_neighbours_info(neigh_address)[neigh_address]["attach_if"]
+    return get_neighbours_info(neigh_address, session=session)[neigh_address]["attach_if"]
```
Comment (vivianQizhu) [inline virttest/utils_net.py]: Better make lines within 79 characters.

**Comment by balamuruhans** [comment]
@vivianQizhu, I have made couple of changes
1. retry arp/ip route to verify IP address for `timeout` because after migration VM would take some time to respond and if there is a race condition where framework checks for arp/ip route before VM starts responding in target could error out the tests unnecessarily, so retry it for timeout and raise would be appropriate.

2. use `timeout` for all the commands executed using `process` or `session`.

Please help to review.

**Comment by vivianQizhu** [comment]
@xiywang Request for your review, especially for the timeout value.

**Comment by xiywang** [comment]
The timeout value looks good to me though it's a bit large. But consider combining with migration, I think it's acceptable.

Acked-by: Xiyue Wang xiywang@redhat.com

**Comment by balamuruhans** [comment]
**Tested the patch:**

_In source host_
```
2019-04-10 02:43:08,730 client           L1126 DEBUG| Sending command: cat /proc/uptime
2019-04-10 02:43:09,061 virsh_migrate_st L0094 INFO | uptime of VM vms1: 153.09

2019-04-10 02:43:09,062 libvirt          L1292 INFO | Check VM network connectivity before migration
2019-04-10 02:43:09,169 virt_vm          L1167 DEBUG| Attempting to log into 'vms1' (timeout 240s)
2019-04-10 02:43:09,171 utils_misc       L0586 DEBUG| Retry verifying IP address (0.000009 secs)
2019-04-10 02:43:09,364 virt_vm          L0876 DEBUG| Found/Verified IP 192.168.122.131 for VM vms1 NIC 0
2019-04-10 02:43:09,366 utils_misc       L0586 DEBUG| Retry verifying IP address (0.000012 secs)
2019-04-10 02:43:11,088 utils_misc       L0586 DEBUG| Retry verifying IP address (0.000016 secs)
2019-04-10 02:43:11,285 utils_net        L0825 INFO | The command of Ping is: ping 192.168.122.131  -c 10
2019-04-10 02:43:11,375 client           L0605 DEBUG| PING 192.168.122.131 (192.168.122.131) 56(84) bytes of data.
2019-04-10 02:43:11,375 client           L0605 DEBUG| 64 bytes from 192.168.122.131: icmp_seq=1 ttl=64 time=0.817 ms
2019-04-10 02:43:12,455 client           L0605 DEBUG| 64 bytes from 192.168.122.131: icmp_seq=2 ttl=64 time=0.939 ms
2019-04-10 02:43:13,456 client           L0605 DEBUG| 64 bytes from 192.168.122.131: icmp_seq=3 ttl=64 time=0.880 ms
2019-04-10 02:43:14,535 client           L0605 DEBUG| 64 bytes from 192.168.122.131: icmp_seq=4 ttl=64 time=0.909 ms
2019-04-10 02:43:15,536 client           L0605 DEBUG| 64 bytes from 192.168.122.131: icmp_seq=5 ttl=64 time=0.934 ms
2019-04-10 02:43:16,538 client           L0605 DEBUG| 64 bytes from 192.168.122.131: icmp_seq=6 ttl=64 time=1.01 ms
2019-04-10 02:43:17,538 client           L0605 DEBUG| 64 bytes from 192.168.122.131: icmp_seq=7 ttl=64 time=0.765 ms
2019-04-10 02:43:18,617 client           L0605 DEBUG| 64 bytes from 192.168.122.131: icmp_seq=8 ttl=64 time=1.17 ms
2019-04-10 02:43:19,617 client           L0605 DEBUG| 64 bytes from 192.168.122.131: icmp_seq=9 ttl=64 time=0.777 ms
2019-04-10 02:43:20,696 client           L0605 DEBUG| 64 bytes from 192.168.122.131: icmp_seq=10 ttl=64 time=0.891 ms
2019-04-10 02:43:20,699 client           L0605 DEBUG|
2019-04-10 02:43:20,701 client           L0605 DEBUG| --- 192.168.122.131 ping statistics ---
2019-04-10 02:43:20,703 client           L0605 DEBUG| 10 packets transmitted, 10 received, 0% packet loss, time 9319ms
```

_In target host_
```
2019-04-10 02:59:26,751 client           L1126 DEBUG| Sending command: cat /proc/uptime
2019-04-10 02:59:26,865 virsh_migrate_st L0050 INFO | uptime of migrated VM vms1: 1130.88
2019-04-10 02:59:28,072 virsh_migrate_st L0057 INFO | Guest state is 'running' at destination is as expected
2019-04-10 02:59:28,072 remote           L0397 DEBUG| Attempting to log into 9.40.193.215:22 using ssh (timeout 240s)
2019-04-10 02:59:29,635 utils_misc       L0586 DEBUG| Retry verifying IP address (0.000009 secs)
2019-04-10 02:59:29,636 client           L1162 DEBUG| Sending command (safe): cat /proc/net/arp
2019-04-10 02:59:29,741 libvirt          L1290 INFO | Check VM network connectivity after migrating
2019-04-10 02:59:29,741 client           L1162 DEBUG| Sending command (safe): echo %OS%
2019-04-10 02:59:29,844 client           L1126 DEBUG| Sending command: ping 192.168.122.131  -c 10
2019-04-10 02:59:29,951 client           L0852 DEBUG| PING 192.168.122.131 (192.168.122.131) 56(84) bytes of data.
2019-04-10 02:59:29,951 client           L0852 DEBUG| 64 bytes from 192.168.122.131: icmp_seq=1 ttl=64 time=0.385 ms
2019-04-10 02:59:31,027 client           L0852 DEBUG| 64 bytes from 192.168.122.131: icmp_seq=2 ttl=64 time=0.373 ms
2019-04-10 02:59:32,066 client           L0852 DEBUG| 64 bytes from 192.168.122.131: icmp_seq=3 ttl=64 time=0.359 ms
2019-04-10 02:59:33,106 client           L0852 DEBUG| 64 bytes from 192.168.122.131: icmp_seq=4 ttl=64 time=0.363 ms
2019-04-10 02:59:34,136 client           L0852 DEBUG| 64 bytes from 192.168.122.131: icmp_seq=5 ttl=64 time=0.320 ms
2019-04-10 02:59:35,187 client           L0852 DEBUG| 64 bytes from 192.168.122.131: icmp_seq=6 ttl=64 time=0.350 ms
2019-04-10 02:59:36,217 client           L0852 DEBUG| 64 bytes from 192.168.122.131: icmp_seq=7 ttl=64 time=0.308 ms
2019-04-10 02:59:37,258 client           L0852 DEBUG| 64 bytes from 192.168.122.131: icmp_seq=8 ttl=64 time=0.378 ms
2019-04-10 02:59:38,308 client           L0852 DEBUG| 64 bytes from 192.168.122.131: icmp_seq=9 ttl=64 time=1.70 ms
2019-04-10 02:59:39,310 client           L0852 DEBUG| 64 bytes from 192.168.122.131: icmp_seq=10 ttl=64 time=0.295 ms
2019-04-10 02:59:39,310 client           L0852 DEBUG|
2019-04-10 02:59:39,311 client           L0852 DEBUG| --- 192.168.122.131 ping statistics ---
2019-04-10 02:59:39,311 client           L0852 DEBUG| 10 packets transmitted, 10 received, 0% packet loss, time 9357ms
2019-04-10 02:59:39,311 client           L0852 DEBUG| rtt min/avg/max/mdev = 0.295/0.483/1.705/0.408 ms

```

**Comment by balamuruhans** [comment]
Thank you @vivianQizhu @xiywang , let us merge it.

**Pull request merged**